### PR TITLE
Added Healthchecks for services

### LINKED ISSUE #3: [stencil] Out of the box health checks
## ToDo:
- [ ] Add support for having a health-check end point out of the box for each controller that is created.
       - [ ] DB

### PR DESCRIPTION
**Pull request by RyanWalker277**
Fixes #3

### CODE REVIEW

File: sample/01-all-features/src/health/health.controller.spec.ts
Comment (techsavvyash) [inline sample/01-all-features/src/health/health.controller.spec.ts:1]: let's add tests or remove this file completely.

File: sample/01-all-features/src/health/health.controller.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
+import { Controller, Get } from '@nestjs/common';
+import { HealthCheck, HealthCheckResult, HealthCheckService } from '@nestjs/terminus';
+import { getConnection, createConnection, Connection } from 'typeorm';
+import axios from 'axios';
+import Redis from 'ioredis';
+import createRedisHealthcheck from 'redis-healthcheck';
+
+interface DatabaseConnectionDetails {
```
Comment (techsavvyash) [inline sample/01-all-features/src/health/health.controller.ts:8]: let's abstract out these interfaces as export from a file called `health.interface.ts`

File: sample/01-all-features/src/health/health.controller.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        healthEndpoint: process.env.FUSIONAUTH_BASE_URL + '/api/status',
+      }),
+      async () => this.checkRedis({
+        host: process.env.REDIS_HOST || 'localhost',
+        port: parseInt(process.env.REDIS_PORT || '6379', 10),
+      }),
+    ]);
+  }
+
+  private async checkDatabase(connectionDetails: DatabaseConnectionDetails): Promise<any> {
```
Comment (techsavvyash) [inline sample/01-all-features/src/health/health.controller.ts:63]: let's abstract out these `checkDatabase`, `checkFusionAuth` etc functions inside a service file as `health.service.ts`

File: sample/01-all-features/src/health/health.controller.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    try {
+      let connection: Connection | undefined;
+
+      try {
+        connection = getConnection(connectionDetails.name);
+      } catch (error) {
+        connection = await createConnection({
+          name: connectionDetails.name,
```
Comment (techsavvyash) [inline sample/01-all-features/src/health/health.controller.ts:71]: nit: this can replaced with `{ ...connectionDetails }`

**Review by techsavvyash — CHANGES_REQUESTED** [review]
let's also add some sort of config in while registering the `HealthCheckModule` in the `app.module.ts` to define what services to actually check. This can be done by creating a `DynamicModule` refers NestJS docs for that.

**Review by techsavvyash — COMMENTED** [review]
The changes here look good, but where have we added the configurability to dynamically define what all services are we using?

**Comment by RyanWalker277** [comment]
Hey @techsavvyash, pushed the changes, can you review this?

File: sample/01-all-features/package.json [hunk 4]
```diff
@@ -33,17 +33,21 @@
     "@nestjs/passport": "10.0.0",
     "@nestjs/platform-express": "10.0.5",
     "@nestjs/swagger": "7.1.1",
+    "@nestjs/terminus": "^10.2.0",
+    "@nestjs/typeorm": "^10.0.1",
```
Comment (techsavvyash) [inline sample/01-all-features/package.json:37]: why are we using typeorm?

File: sample/01-all-features/package.json [hunk 5]
```diff
@@ -52,14 +56,17 @@
...
+    "redis": "^4.6.12",
+    "redis-healthcheck": "^0.1.1",
     "reflect-metadata": "0.1.13",
     "rimraf": "5.0.1",
     "rxjs": "7.8.1",
     "source-map-support": "0.5.21",
     "swagger-ui-express": "4.6.3",
-    "twitter": "1.7.1"
+    "twitter": "1.7.1",
+    "typeorm": "^0.3.17"
```
Comment (techsavvyash) [inline sample/01-all-features/package.json:69]: why is this required?

File: sample/01-all-features/src/app.module.ts [hunk 6]
```diff
@@ -64,6 +65,7 @@ import { user } from '@techsavvyash/user-service';
     MailerModule,
     HomeModule,
     user.UserModule,
+    HealthModule.forRoot(['postgres', 'shadowpostgres', 'fusionauth', 'redis']),
```
Comment (techsavvyash) [inline sample/01-all-features/src/app.module.ts:68]: can we turn these strings into exported constants?

File: sample/01-all-features/src/health/health.interface.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+export interface DatabaseConnectionDetails {
```
Comment (techsavvyash) [inline sample/01-all-features/src/health/health.interface.ts:1]: can we rename the class to be Postgres specific? 
If this same structure can be used for other SQL based DBs? In that case can we turn this one into a generic object?

File: sample/01-all-features/src/health/health.service.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
+// health.service.ts
+
+import { Injectable } from '@nestjs/common';
+import { getConnection, createConnection, Connection } from 'typeorm';
```
Comment (techsavvyash) [inline sample/01-all-features/src/health/health.service.ts:4]: can we try remove this dependency on typeORM

**Review by techsavvyash — CHANGES_REQUESTED** [review]

File: sample/01-all-features/src/app.module.ts (hunk 6 above)
Comment (RyanWalker277) [reply sample/01-all-features/src/app.module.ts:68]: Instead of exported constants, shouldn't we pick them from the `.env` itself?

Comment (techsavvyash) [reply sample/01-all-features/src/app.module.ts:68]: I think passing in constants would be a better choice. Here's my reasoning:
1. Having exported types will prevent typos in manual typing and this approach will be more receptive to any internal changes to those constants we might have to make.
2. It is not very often the case when a service depending on other services is not requiring all of them to function. In the event that is actually the case, we can have a second level enable/disable flag from the .env which is not required but optional

**Review by techsavvyash — APPROVED** [review]
LGTM 👍🏻
